drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1044 @@
+			ret = spi_nand_conti_read_enable(dev, true);
+			if (ret != 0) {
+				LOG_ERR("SPI NAND Set continuous read enable Failed: %d", ret);
+				return ret;
+			}
+		} else {
+			data->continuous_read = false;
+		}
+	} else {
+		LOG_ERR("ONFI table not found");
+		return EXIT_SUCCESS;
+	}
+
+	acquire_device(dev);
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out1;
+	}
+
+	secur_reg &= ~SPINAND_SECURE_BIT_OTP_EN;
+	ret = spi_nand_set_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, secur_reg);
+	if (ret != 0) {
+		LOG_ERR("set feature failed: %d", ret);
+		goto out1;
+	}
+
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out1;
+	}
+
+	if (secur_reg & SPINAND_SECURE_BIT_OTP_EN) {
+		LOG_ERR("Disable OTP failed: %d\n", secur_reg);
+		ret = -EINVAL;
+	}
+
+out1:
+	release_device(dev);
+	return ret;
+}
+
+/**
+ * @brief Configure the flash
+ *
+ * @param dev The flash device structure
+ * @param info The flash info structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static int spi_nand_configure(const struct device *dev)
+{
+	const struct spi_nand_config *cfg = dev->config;
+
+	uint8_t reg = 0;
+	int ret;
+	/* Validate bus and CS is ready */
+	if (!spi_is_ready_dt(&cfg->spi)) {
+		return -ENODEV;
+	}
+	ret = spi_nand_check_id(dev);
+	if (ret != 0) {
+		LOG_ERR("Check ID failed: ");
+		return -ENODEV;
+	}
+	/* Check for block protect bits that need to be cleared. */
+	acquire_device(dev);
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_BLOCK_PROT, &reg);
+
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out;
+	}
+	/* Only clear if GET_FEATURE worked and something's set. */
+	if (reg & SPINAND_BLOCK_PROT_BIT_BP_MASK) {
+		reg = 0;
+		ret = spi_nand_set_feature(dev, SPI_NAND_FEA_ADDR_BLOCK_PROT, reg);
+		if (ret != 0) {
+			LOG_ERR("set feature failed: %d", ret);
+		}
+	}
+out:
+	release_device(dev);
+
+	if (ret != 0) {
+		return ret;
+	}
+	ret = spi_nand_read_otp_onfi(dev);
+	if (ret != 0) {
+		LOG_ERR("SFDP read failed: %d", ret);
+		return -ENODEV;
+	}
+
+	return EXIT_SUCCESS;
+}
+
+/**
+ * @brief Initialize and configure the flash
+ *
+ * @param name The flash name
+ * @return 0 on success, negative errno code otherwise
+ */
+static int spi_nand_init(const struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		struct spi_nand_data *const driver_data = dev->data;
+
+		k_sem_init(&driver_data->sem, 1, K_SEM_MAX_LIMIT);
+	}
+
+	return spi_nand_configure(dev);
+}
+
+static const struct flash_parameters flash_nand_parameters = {
+	.write_block_size = FLASH_WRITE_SIZE,
+	.erase_value = 0xff,
+};
+
+static const struct flash_parameters *
+flash_nand_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_nand_parameters;
+}
+
+static const struct flash_driver_api spi_nand_api = {
```

Comment:
Driver API needs to be wrapped in macro.

```suggestion
static DEVICE_API(flash, spi_nand_api) = {
```

Reply:
Resolved